docs/rfc/rfc-027-p2p-message-bandwidth-report.md

```diff
@@ -0,0 +1,137 @@
+
+## Discussion
+
+### Overview of Major Bandwidth Usage
+
+To determine which components of Tendermint were consuming the most bandwidth, I gathered prometheus metrics from the [Blockpane][blockpane] validator running on the Osmosis network for several hours.
+The data reveal that three message types account for 98% of the total bandwidth consumed. These message types are as follows:
+
+1. [consensus.BlockPart][block-part-message]
+2. [mempool.Txs][mempool-txs-message]
+3. [consensus.Vote][vote-message]
+
+
+The image below of p2p data collected from the Blockpane validator illustrates the bandwidth consumption of these three message types in both data sending and receiving.
+
+![](images/total-message-send-rate.png)
+![](images/total-message-recv-rate.png)
+
+### Investigation of Message Usage
+
+This section discusses the usage of each of the three highest consumption messages.
+
+#### BlockPart Transmission
+
+Sending `BlockPart` messages consumes the most bandwidth out of all p2p messages types as observed in the Blockpane Osmosis validator.
+In the almost 3 hour observation, the validator sent about 20 gigabytes of `BlockPart` messages.
+
+A block is proposed each round of Tendermint consensus. The paper does not define a specific way that
+the block is to be transmitted, just that all participants will receive it via a gossip network.
+
+The Go implementation of Tendermint transmits the block in 'parts'. It serializes
+the block to wire-format proto and splits this byte representation into a set of
+4 kilobyte arrays and sends these arrays to its peers, each in a separate message.
+
+The logic for sending `BlockPart` messages resides in the code for the [consensus.Reactor][gossip-data-routine].
+The consensus reactor starts a new `gossipDataRoutine` for each peer it connects to.
+This routine repeatedly picks a part of the block that Tendermint believes the peer does not know about yet and gossips it to the peer.
+The set of `BlockParts` that Tendermint considers its peer as having is only updated in one of four ways:
+
+1. [We receive a block part from the peer][block-part-receive].
+2. [We send][block-part-send-1] [the peer a block part][block-part-send-2].
+3. Our peer tells us about the parts they have block [via `NewValidBlock` messages][new-valid-block-message-send]. This message is only sent when the validator has a quorum of prevotes or precommits for a block.
+4. Our peer tells us they have entered a new round [via a `NewRoundStep` message][new-round-step-message-send].
+
+Each node receives block parts from all of its peers. The particular block part to send at any given time is randomly selected from the set of parts that the peer node is not yet known to have. Given that these are the only times that Tendermint learns of its peers' block parts, it's very likely that a node has an incomplete understanding of its peers' block parts and is transmitting block parts to a peer that the peer has received from some other node.
```

Comment:
I would suggest breaking this line into multiple lines, for better versioning.

I think the point here is that we potentially will exchange (send or receive) all block parts with all peers, as the mechanism 1 and 2 should prevent sending the same block part more than once through a link (in any direction).

Reply:
Done.